File: go/internal/resolver/crdt_resolver.go
package resolver

import (
	"time"

	"github.com/knirv/knirvbase/internal/clock"
	"github.com/knirv/knirvbase/internal/types"
)

// ResolveConflict applies LWW + vector-clock tie-breaking
func ResolveConflict(local, remote *types.DistributedDocument) *types.DistributedDocument {
	if remote == nil {
		return local
	}
	if local == nil {
		return remote
	}

	// Deletion handling
	if remote.Deleted && !local.Deleted {
		comp := clock.Compare(local.Vector, remote.Vector)
		if comp == clock.Before || comp == clock.Concurrent {
			return remote
		}
		return local
	}
	if local.Deleted && !remote.Deleted {
		comp := clock.Compare(remote.Vector, local.Vector)
		if comp == clock.Before || comp == clock.Concurrent {
			return local
		}
		return remote
	}

	switch clock.Compare(local.Vector, remote.Vector) {
	case clock.After:
		return local
	case clock.Before:
		return remote
	case clock.Equal:
		return local
	case clock.Concurrent:
		// Use timestamp, then peer id for deterministic ordering
		if local.Timestamp > remote.Timestamp {
			return mergeDocuments(local, remote)
		} else if local.Timestamp < remote.Timestamp {
			return mergeDocuments(remote, local)
		}
		if local.PeerID >= remote.PeerID {
			return mergeDocuments(local, remote)
		}
		return mergeDocuments(remote, local)
	}
	return local
}

func mergeDocuments(winner, loser *types.DistributedDocument) *types.DistributedDocument {
	merged := *winner // shallow copy
	merged.Vector = clock.Merge(winner.Vector, loser.Vector)

	// Merge non-conflicting fields from loser payload
	if merged.Payload == nil {
		merged.Payload = make(map[string]interface{})
	}
	for k, v := range loser.Payload {
		if _, ok := merged.Payload[k]; !ok {
			merged.Payload[k] = v
		}
	}
	return &merged
}

// ApplyOperation applies a CRDT operation to a document (insert/update/delete)
// Note: Documents marked with Stage == "post-pending" are treated as a local intent
// and are not emitted as regular CRDT operations until they are explicitly posted as
// KNIRVGRAPH transactions during the sync flow.
func ApplyOperation(doc *types.DistributedDocument, op types.CRDTOperation) *types.DistributedDocument {
	switch op.Type {
	case types.OpInsert, types.OpUpdate:
		if doc == nil {
			if op.Data == nil {
				return nil
			}
			copy := *op.Data
			copy.Vector = clock.Clone(op.Vector)
			copy.Timestamp = op.Timestamp
			copy.PeerID = op.PeerID
			return &copy
		}

		comp := clock.Compare(doc.Vector, op.Vector)
		if comp == clock.Before || comp == clock.Concurrent {
			// Merge fields
			if doc.Payload == nil {
				doc.Payload = make(map[string]interface{})
			}
			for k, v := range op.Data.Payload {
				doc.Payload[k] = v
			}
			doc.Vector = clock.Merge(doc.Vector, op.Vector)
			if op.Timestamp > doc.Timestamp {
				doc.Timestamp = op.Timestamp
			}
		}
		return doc

	case types.OpDelete:
		if doc == nil {
			return nil
		}
		comp := clock.Compare(doc.Vector, op.Vector)
		if comp == clock.Before || comp == clock.Concurrent {
			doc.Deleted = true
			doc.Vector = clock.Merge(doc.Vector, op.Vector)
			if op.Timestamp > doc.Timestamp {
				doc.Timestamp = op.Timestamp
			}
		}
		return doc
	default:
		return doc
	}
}

// ToDistributed converts a simple map payload to DistributedDocument
func ToDistributed(payload map[string]interface{}, peerID string) *types.DistributedDocument {
	v := clock.NewVectorClock()
	v[peerID] = 1
	return &types.DistributedDocument{
		ID:        payload["id"].(string),
		Payload:   cloneMap(payload),
		Vector:    v,
		Timestamp: time.Now().UnixMilli(),
		PeerID:    peerID,
	}
}

// ToRegular converts a DistributedDocument to a regular map representation
func ToRegular(doc *types.DistributedDocument) map[string]interface{} {
	if doc == nil {
		return nil
	}
	return cloneMap(doc.Payload)
}

// cloneMap creates a shallow copy of a map
func cloneMap(src map[string]interface{}) map[string]interface{} {
	if src == nil {
		return nil
	}
	dst := make(map[string]interface{}, len(src))
	for k, v := range src {
		dst[k] = v
	}
	return dst
}

File: ts/src/components/resolver/crdt_resolver.ts
import { VectorClock, compare, merge, clone, ComparisonResult } from '../clock/vector_clock';
import { DistributedDocument, CRDTOperation, OperationType } from '../types/types';

// ResolveConflict applies LWW + vector-clock tie-breaking
export function ResolveConflict(local: DistributedDocument | null, remote: DistributedDocument | null): DistributedDocument | null {
  if (!remote) return local;
  if (!local) return remote;

  // Deletion handling
  if (remote._deleted && !local._deleted) {
    const comp = compare(local._vector, remote._vector);
    if (comp === ComparisonResult.Before || comp === ComparisonResult.Concurrent) {
      return remote;
    }
    return local;
  }
  if (local._deleted && !remote._deleted) {
    const comp = compare(remote._vector, local._vector);
    if (comp === ComparisonResult.Before || comp === ComparisonResult.Concurrent) {
      return local;
    }
    return remote;
  }

  switch (compare(local._vector, remote._vector)) {
    case ComparisonResult.After:
      return local;
    case ComparisonResult.Before:
      return remote;
    case ComparisonResult.Equal:
      return local;
    case ComparisonResult.Concurrent:
      // Use timestamp, then peer id for deterministic ordering
      if (local._timestamp > remote._timestamp) {
        return mergeDocuments(local, remote);
      } else if (local._timestamp < remote._timestamp) {
        return mergeDocuments(remote, local);
      }
      if (local._peerId >= remote._peerId) {
        return mergeDocuments(local, remote);
      }
      return mergeDocuments(remote, local);
  }
  return local;
}

function mergeDocuments(winner: DistributedDocument, loser: DistributedDocument): DistributedDocument {
  const merged: DistributedDocument = { ...winner };
  merged._vector = merge(winner._vector, loser._vector);

  // Merge non-conflicting fields from loser payload
  if (!merged.payload) {
    merged.payload = {};
  }
  for (const [k, v] of Object.entries(loser.payload || {})) {
    if (!(k in merged.payload)) {
      merged.payload[k] = v;
    }
  }
  return merged;
}

// ApplyOperation applies a CRDT operation to a document (insert/update/delete)
export function ApplyOperation(doc: DistributedDocument | null, op: CRDTOperation): DistributedDocument | null {
  switch (op.type) {
    case OperationType.Insert:
    case OperationType.Update:
      if (!doc) {
        if (!op.data) return null;
        const copy: DistributedDocument = { ...op.data };
        copy._vector = clone(op.vector);
        copy._timestamp = op.timestamp;
        copy._peerId = op.peerId;
        return copy;
      }

      const comp = compare(doc._vector, op.vector);
      if (comp === ComparisonResult.Before || comp === ComparisonResult.Concurrent) {
        // Merge fields
        if (!doc.payload) doc.payload = {};
        for (const [k, v] of Object.entries(op.data?.payload || {})) {
          doc.payload[k] = v;
        }
        doc._vector = merge(doc._vector, op.vector);
        if (op.timestamp > doc._timestamp) {
          doc._timestamp = op.timestamp;
        }
      }
      return doc;

    case OperationType.Delete:
      if (!doc) return null;
      const compDel = compare(doc._vector, op.vector);
      if (compDel === ComparisonResult.Before || compDel === ComparisonResult.Concurrent) {
        doc._deleted = true;
        doc._vector = merge(doc._vector, op.vector);
        if (op.timestamp > doc._timestamp) {
          doc._timestamp = op.timestamp;
        }
      }
      return doc;
    default:
      return doc;
  }
}

// ToDistributed converts a simple map payload to DistributedDocument
export function ToDistributed(payload: Record<string, any>, peerID: string): DistributedDocument {
  const v: VectorClock = {};
  v[peerID] = 1;
  return {
    id: payload.id,
    entryType: payload.entryType,
    payload: cloneMap(payload),
    _vector: v,
    _timestamp: Date.now(),
    _peerId: peerID,
  };
}

// ToRegular converts a DistributedDocument to a regular map representation
export function ToRegular(doc: DistributedDocument | null): Record<string, any> | null {
  if (!doc) return null;
  return cloneMap(doc.payload) || null;
}

// cloneMap creates a shallow copy of a map
function cloneMap(src: Record<string, any> | undefined): Record<string, any> | undefined {
  if (!src) return undefined;
  return { ...src };
}